### PR TITLE
Hardening Committee Cache for Runtime

### LINKED ISSUE #4083: Beacon Client Error: Could not receive attestation in chain service
I noticed the following error repeatedly, starting from slot 73579:

```
[2019-11-21 21:44:53] ERROR blockchain: Could not receive attestation in chain service error=attestation aggregation signature did not verify
could not verify indexed attestation
github.com/prysmaticlabs/prysm/beacon-chain/blockchain/forkchoice.(*Store).verifyAttestation
        beacon-chain/blockchain/forkchoice/process_attestation.go:170
github.com/prysmaticlabs/prysm/beacon-chain/blockchain/forkchoice.(*Store).OnAttestation
        beacon-chain/blockchain/forkchoice/process_attestation.go:90
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).ReceiveAttestationNoPubsub
        beacon-chain/blockchain/receive_attestation.go:30
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).processAttestation
        beacon-chain/blockchain/receive_attestation.go:68
runtime.goexit
        GOROOT/src/runtime/asm_amd64.s:1357
could not process attestation from fork choice service
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).ReceiveAttestationNoPubsub
        beacon-chain/blockchain/receive_attestation.go:31
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).processAttestation
        beacon-chain/blockchain/receive_attestation.go:68
runtime.goexit
        GOROOT/src/runtime/asm_amd64.s:1357
```

As far as I can tell the client then continued to run just fine from slot 73585 onwards.

### LINKED ISSUE #4248: Committee cache handles N+1 fork scenario
Currently the committee list cache does not handle reorg which invalidates cached shuffled indices at epoch N+1. The program is outlined below:
* At the start of epoch N, a node should be able to cache shuffling for N and N + 1
* Such shuffling is determined by epoch N and randao seed at the start of N
* If the chain reorg such that randao at the start of N is different, then the shuffled indices for N + 1 would be invalidated
* The cache for N + 1 is wrt to epoch N and randao at the start of N

Solution:
We should keep old value around if such a reorg because that shows some serious network instability and a node might have multiple branches that you need to be able to validate different shufflings against. Update the cache to be keyed by epoch and epoch boundary block info

### PR DESCRIPTION
**Pull request by terencechain**
This fixes #4248 and fixes #4083 

This updates runtime committee cache to use RANDAO seed as key (was epoch). It added a fall back to no-cache sig verify when cache sig fails to verify
Change list:
- Use RANDAO seed as key
- Added fallback `convertToIndexed` to use when sig fails to verify
- Added reattempt to verify signature without cache
- Added attestation-failed-to-verify metric
- Removed unused cache getters `Epochs()`, `EpochInCache()` and `CommitteeCountPerSlot()`
- Tested both run time and local set up
- ~~Updates cache upon miss, not at predetermined boundary slot~~

### CODE REVIEW

File: beacon-chain/cache/committee.go [hunk 1]
```diff
@@ -18,9 +18,10 @@ var (
 	// a Committee struct.
 	ErrNotCommittee = errors.New("object is not a committee struct")
 
-	// maxShuffledIndicesSize defines the max number of shuffled indices list can cache.
-	// 3 for previous, current epoch and next epoch.
-	maxShuffledIndicesSize = 3
+	// maxCommitteeSize defines the max number of shuffled committees one can cache
+	// on per epoch basis.
+	// Due to reorgs, it's good to keep the old cache around for quickly switch over.
+	maxCommitteeSize = 10
```
Comment (0xKiwi) [inline beacon-chain/cache/committee.go]: This name is a little vague, this is tracking how many committees? Not the size of the committees, right?

File: beacon-chain/core/helpers/validators.go [hunk 2]
```diff
@@ -58,7 +58,11 @@ func IsSlashableValidator(validator *ethpb.Validator, epoch uint64) bool {
 //    return [ValidatorIndex(i) for i, v in enumerate(state.validators) if is_active_validator(v, epoch)]
 func ActiveValidatorIndices(state *pb.BeaconState, epoch uint64) ([]uint64, error) {
 	if featureconfig.Get().EnableNewCache {
-		activeIndices, err := committeeCache.ActiveIndices(epoch)
+		seed, err := Seed(state, epoch, params.BeaconConfig().DomainBeaconAttester)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not get seed")
+		}
+		activeIndices, err := committeeCache.ActiveIndices(epoch, seed)
```
Comment (prestonvanloon) [inline beacon-chain/core/helpers/validators.go]: This method never sets the active indices on a cache miss.

File: beacon-chain/cache/committee.go [hunk 3]
```diff
@@ -218,3 +156,7 @@ func startEndIndices(c *Committee, index uint64) (uint64, uint64) {
 
 	return start, end
 }
+
+func key(slot uint64, seed [32]byte) string {
```
Comment (prestonvanloon) [inline beacon-chain/cache/committee.go]: Why is the seed a key? Why the slot when everything else says epoch?

Can you please add commentary on why this is a good key function for committee caching?

File: beacon-chain/cache/committee_test.go [hunk 4]
```diff
@@ -69,111 +71,56 @@ func TestCommitteeCache_CommitteesByEpoch(t *testing.T) {
...
+	seed := [32]byte{'A'}
 
-	item := &Committee{Epoch: 1, Committee: []uint64{1, 2, 3, 4, 5, 6}}
-	indices, err := cache.ActiveIndices(1)
-	if err != nil {
-		t.Fatal(err)
-	}
-	if indices != nil {
-		t.Error("Expected committee count not to exist in empty cache")
+	// Should rotate out all the epochs except 190 to 199
```
Comment (0xKiwi) [inline beacon-chain/cache/committee_test.go]:
```suggestion
	// Should rotate out all the epochs except 190 through 199.
```

File: beacon-chain/core/blocks/block_operations.go [hunk 5]
```diff
@@ -812,6 +812,7 @@ func VerifyIndexedAttestation(ctx context.Context, beaconState *pb.BeaconState,
 	hasVotes := len(custodyBit0Indices) > 0 || len(custodyBit1Indices) > 0
 
 	if hasVotes && !sig.VerifyAggregate(pubkeys, msgs, domain) {
+		fmt.Println()
```
Comment (prestonvanloon) [inline beacon-chain/core/blocks/block_operations.go]: Remove please

Comment (terencechain) [reply beacon-chain/core/blocks/block_operations.go]: PR is not ready for review yet. See wip

File: beacon-chain/cache/committee.go (hunk 3 above)
Comment (terencechain) [reply beacon-chain/cache/committee.go]: Using seed and added a comment 👍

File: beacon-chain/blockchain/forkchoice/metrics.go [hunk 6]
```diff
@@ -57,6 +57,10 @@ var (
 		Name: "beacon_current_validators",
 		Help: "Number of status=pending|active|exited|withdrawable validators in current epoch",
 	})
+	sigFailsToVerify = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "att_signature_failed_to_verify",
+		Help: "Number of attestation signature failed to verify",
```
Comment (0xKiwi) [inline beacon-chain/blockchain/forkchoice/metrics.go]:
```suggestion
		Help: "Number of attestation signatures that failed to verify",
```

File: beacon-chain/cache/committee.go [hunk 7]
```diff
@@ -18,9 +17,9 @@ var (
 	// a Committee struct.
 	ErrNotCommittee = errors.New("object is not a committee struct")
 
-	// maxShuffledIndicesSize defines the max number of shuffled indices list can cache.
-	// 3 for previous, current epoch and next epoch.
-	maxShuffledIndicesSize = 3
+	// maxCommitteesSize defines the max number of shuffled committees on per randao basis can cache.
+	// Due to reorgs, it's good to keep the old cache around for quickly switch over.
+	maxCommitteesSize = 10
```
Comment (shayzluf) [inline beacon-chain/cache/committee.go]:
```suggestion
	maxCommitteesCacheSize = 10
```

Comment (terencechain) [reply beacon-chain/cache/committee.go]: Thanks!

File: beacon-chain/cache/committee.go [hunk 8]
```diff
@@ -98,100 +97,36 @@ func (c *CommitteeCache) ShuffledIndices(slot uint64, index uint64) ([]uint64, e
...
 	start, end := startEndIndices(item, indexOffSet)
-	return item.Committee[start:end], nil
+
+	return item.ShuffledIndices[start:end], nil
 }
 
 // AddCommitteeShuffledList adds Committee shuffled list object to the cache. T
 // his method also trims the least recently list if the cache size has ready the max cache size limit.
-func (c *CommitteeCache) AddCommitteeShuffledList(committee *Committee) error {
+func (c *CommitteeCache) AddCommitteeShuffledList(committee *Committees) error {
```
Comment (shayzluf) [inline beacon-chain/cache/committee.go]:
```suggestion
func (c *CommitteeCache) AddCommitteeShuffledList(committees *Committees) error {
```

File: beacon-chain/cache/committee.go [hunk 9]
```diff
@@ -98,100 +97,36 @@ func (c *CommitteeCache) ShuffledIndices(slot uint64, index uint64) ([]uint64, e
...
 	if !featureconfig.Get().EnableShuffledIndexCache && !featureconfig.Get().EnableNewCache {
 		return nil
 	}
 	c.lock.Lock()
 	defer c.lock.Unlock()
+
 	if err := c.CommitteeCache.AddIfNotPresent(committee); err != nil {
```
Comment (shayzluf) [inline beacon-chain/cache/committee.go]:
```suggestion
	if err := c.CommitteeCache.AddIfNotPresent(committees); err != nil {
```

File: beacon-chain/blockchain/forkchoice/metrics.go [hunk 10]
```diff
@@ -57,6 +57,10 @@ var (
 		Name: "beacon_current_validators",
 		Help: "Number of status=pending|active|exited|withdrawable validators in current epoch",
 	})
+	sigFailsToVerify = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "att_signature_failed_to_verify",
+		Help: "Number of attestation signatures that failed to verify",
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/forkchoice/metrics.go]: Can you clarify the intention of this metric? What I see in the code is that you increment this when a signature failed to verify from the cache and then you manually recheck it and then it succeeded to verify

File: beacon-chain/blockchain/forkchoice/process_block.go [hunk 11]
```diff
@@ -135,13 +135,6 @@ func (s *Store) OnBlock(ctx context.Context, b *ethpb.BeaconBlock) error {
 	if helpers.IsEpochStart(postState.Slot) {
 		logEpochData(postState)
 		reportEpochMetrics(postState)
-
-		// Update committee shuffled indices at the end of every epoch
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/forkchoice/process_block.go]: Why remove this?

File: beacon-chain/cache/committee.go [hunk 12]
```diff
@@ -203,18 +138,22 @@ func (c *CommitteeCache) ActiveIndices(epoch uint64) ([]uint64, error) {
...
-	validatorCount := uint64(len(c.Committee))
+func startEndIndices(c *Committees, index uint64) (uint64, uint64) {
+	validatorCount := uint64(len(c.ShuffledIndices))
 	start := sliceutil.SplitOffset(validatorCount, c.CommitteeCount, index)
 	end := sliceutil.SplitOffset(validatorCount, c.CommitteeCount, index+1)
-
 	return start, end
 }
+
+// Using seed as source for key to handle reorgs in the same epoch.
```
Comment (prestonvanloon) [inline beacon-chain/cache/committee.go]: Why is a seed a good key though? How is it used to compute assignments?

File: beacon-chain/blockchain/forkchoice/metrics.go (hunk 10 above)
Comment (terencechain) [reply beacon-chain/blockchain/forkchoice/metrics.go]: It's manually recheck with the cache off

Clarified in comments

File: beacon-chain/blockchain/forkchoice/process_block.go (hunk 11 above)
Comment (terencechain) [reply beacon-chain/blockchain/forkchoice/process_block.go]: The seed->committees updates upon cache miss, not at predetermined boundary slot.

This design was inefficient and no longer needed

Comment (prestonvanloon) [reply beacon-chain/blockchain/forkchoice/process_block.go]: Why is it inefficient? Doesn't it make sense to prepopulate this cache before it's used?

Please update PR description, this change wasn't mentioned.

File: beacon-chain/cache/committee.go (hunk 12 above)
Comment (terencechain) [reply beacon-chain/cache/committee.go]: Added comment

**Comment by terencechain** [comment]
@prestonvanloon It's ready after syncing with #4284

File: beacon-chain/cache/committee.go [hunk 13]
```diff
@@ -18,9 +17,9 @@ var (
 	// a Committee struct.
 	ErrNotCommittee = errors.New("object is not a committee struct")
 
-	// maxShuffledIndicesSize defines the max number of shuffled indices list can cache.
-	// 3 for previous, current epoch and next epoch.
-	maxShuffledIndicesSize = 3
+	// maxCommitteesCacheSize defines the max number of shuffled committees on per randao basis can cache.
+	// Due to reorgs, it's good to keep the old cache around for quickly switch over.
+	maxCommitteesCacheSize = 10
```
Comment (rauljordan) [inline beacon-chain/cache/committee.go]: why 10? can it be explained in the comment?

**Pull request merged**